### PR TITLE
Introduce vstorage work scheme

### PR DESCRIPTION
**Pull request by aburluka**
Signed-off-by: Alexander Burluka <aburluka@virtuozzo.com>

### CODE REVIEW

**Review by dmission — APPROVED** [review]

File: vzstorage-pd/vendor/github.com/avagin/ploop-flexvol/volume/volume.go [hunk 1]
```diff
@@ -1,95 +0,0 @@
...
-import (
-	"fmt"
-	"os"
-	"os/exec"
-
-	"github.com/dustin/go-humanize"
-	"github.com/kolyshkin/goploop-cli"
-)
-
-func Create(options map[string]string) error {
```
Comment (avagin) [inline vzstorage-pd/vendor/github.com/avagin/ploop-flexvol/volume/volume.go:12]: you should create a separate patch to move code from this file. Separate each logical change into a separate patch.

File: vzstorage-pd/virtuozzo-provisioner.go [hunk 2]
```diff
@@ -139,6 +254,15 @@ func (p *vzFSProvisioner) Delete(volume *v1.PersistentVolume) error {
 	}
 
 	options := volume.Spec.PersistentVolumeSource.FlexVolume.Options
+
+	secret := map[string][]byte{
+		"clusterName":     []byte(options["clusterName"]),
```
Comment (avagin) [inline vzstorage-pd/virtuozzo-provisioner.go]: Are you sure that options contain clasterName and clusterPassword? Who added them to options? How does kube describe volume show them?

File: vzstorage-pd/virtuozzo-provisioner.go [hunk 3]
```diff
@@ -83,15 +188,25 @@ func (p *vzFSProvisioner) Provision(options controller.VolumeOptions) (*v1.Persi
...
+	storage_class_options["volumeId"] = share
+	storage_class_options["size"] = fmt.Sprintf("%d", z)
 
-	if err := volume.Create(ploop_options); err != nil {
+	secret, err := p.client.Core().Secrets(storage_class_options["secretNamespace"]).Get(storage_class_options["secretName"])
+	if err != nil {
+		return nil, err
+	}
+
+	if err := initVstorage(storage_class_options, secret.Data); err != nil {
```
Comment (avagin) [inline vzstorage-pd/virtuozzo-provisioner.go]: I think we don't need to mount a vstorage each time. Pls, talk with @dmission, he is solving a similar problem for ploop-flexvol

File: vzstorage-pd/virtuozzo-provisioner.go [hunk 4]
```diff
@@ -67,14 +71,115 @@ func newVzFSProvisioner(client kubernetes.Interface) controller.Provisioner {
...
+
 // Provision creates a storage asset and returns a PV object representing it.
 func (p *vzFSProvisioner) Provision(options controller.VolumeOptions) (*v1.PersistentVolume, error) {
 	var (
 		capacity resource.Quantity
 	)
 
 	capacity = options.PVC.Spec.Resources.Requests[v1.ResourceName(v1.ResourceStorage)]
-	bytes := capacity.Value()
+	z := capacity.Value()
```
Comment (avagin) [inline vzstorage-pd/virtuozzo-provisioner.go]: Could you use more infomative names;)

File: vzstorage-pd/class.yaml [hunk 5]
```diff
@@ -4,4 +4,6 @@ metadata:
   name: virtuozzo-storage
 provisioner: kubernetes.io/virtuozzo-storage
 parameters:
-        volumePath: "/mnt/vstorage/kube/"
+  volumePath: "/mnt/vstorage/kube/"
+  secretNamespace: "default"
```
Comment (dmission) [inline vzstorage-pd/class.yaml]: Please change to 'kube-system' - it is a proper namespace for system-wide services. 'default' is accessible by any user.

**Review by dmission — CHANGES_REQUESTED** [review]

**Pull request merged**